packages/pigeon/README.md

```diff
@@ -1,13 +1,354 @@
-It also improves efficiency over common method channel patterns. Most importantly though,
-it removes the need to write custom platform channel code, since pigeon generates it for you.
+Pigeon works by reading special files which by convention are placed in a separate
+folder, `/pigeons`, alongside your `/lib` folder. You can put all of your definitions
+in a single file (e.g., `/pigeons/messages.dart`), or you can divide your methods
+and classes into separate files as you wish. After reading these files, Pigeon
+generates:
+
+ * New Dart source files within your `/lib` folder at the specific paths you
+specify, and
+ * New native source files with matching code within your native projects at the
+ specific paths you specify
+
+Internally, the generated code uses `MethodChannel`s to communicate across Flutter's
+UI thread and the Platform thread where your host app initially launched. The value
+in Pigeon comes from automatically keeping this unpleasant boilerplate in sync and
+efficiently marshalling data between languages.
+
+The generated code can either flow from Dart to native code, or native code back
+to Dart. Generated code on the receiving end uses interfaces or abstract classes,
+allowing you to provide implementations in concrete classes.
+
+Pigeon works in both complete Flutter apps and hybrid apps using the add-to-app paradigm.
+
+## Quickstart
+
+### Installation
+
+Begin by adding Pigeon to your Dart project's `pubspec.yaml` file:
+
+```sh
+$ flutter pub add pigeon --dev
+$ flutter pub get
+```
+
+### Setup
+
+To specify what code Pigeon should generate, create a `/pigeons` directory and add
+a `.dart` configuration file. This guide will place all such configurations within
+`/pigeons/messages.dart`, but you are free to choose any file name(s) you like.
+
+Begin by instantiating a `PigeonOptions` object, wrapped in the `@ConfigurePigeon`
+decorator. Later, you will pass named parameters to your `PigeonOptions` object
+to specify your desired behavior.
+
+```dart
+// pigeons/messages.dart
+
+import 'package:pigeon/pigeon.dart';
+
+@ConfigurePigeon(
+  PigeonOptions()
+);
+```
+
+### Define your messages
+
+The next step to use Pigeon is to define the data structures Dart will exchange
+with native code. You do this by writing plain Dart enums or classes in your
+`/pigeons` files. These messages should only contain simple constructors and direct
+attributes. Methods, factory constructors, constructor bodies, and computed
+properties are all not allowed.
+
+```dart
+// pigeons/messages.dart
+
+enum Code { a, b }
+
+class MessageData {
+  MessageData({required this.code, required this.data});
+  String? name;
+  String? description;
+  Code code;
+  Map<String, String> data;
+}
+```
+
+No extra steps are necessary to register these classes - their inclusion
+in a `pigeons/*.dart` ensures Pigeon will generate matching Dart and native
+implementations.
+
+### Define which methods to expose
```

Comment:
It's not clear to me how this step is different from the previous step.

Reply:
> `Define your messages` intends to describe defining the data structures that will be passed back and forth; whereas `Define which methods to expose` is about the `HostApi()` and `FlutterApi()` decorators which expose a class with methods.

Many uses of Pigeon don't even require data structures; I don't see these are separate steps (and if they were, certainly wouldn't put defining data structures *before* figuring out what the methods are going to be that would need them).